### DISCUSSION TITLE
break received on communication channel

### DISCUSSION THREAD

**Discussion by aoticombr** [sijms/go-ora#416] [Q&A] [answered]
when executing an insert with return of value where everything is done by hand of err

INSERT INTO MARCAS
(CDMARCAS,MARCAS,ATIVO) ").
 values (null,'xxx1','S') returning CDMARCAS into :id_marca

ORA-03111: break received on communication channel

**Comment by aoticombr** [comment]
if i run it like this it works
begin INSERT INTO BRANDS(BRANDS,ACTIVE) VALUES('xxx1','S') RETURNING CDBRANDS INTO :p3; end;

**Comment by aoticombr** [comment]
or if I put a parameter at least from input it also works
var int brand_id
sqlStatement := `INSERT INTO TAGS(TAGS,ACTIVE) VALUES('xxx1','S') RETURNING CDTAGS INTO :p3`
stmt, err := db.PrepareContext(context.Background(), sqlStatement)
if err != nil {
fmt.Println(err)
}
defer stmt.Close()
if _, err := stmt.ExecContext(context.Background(), sql.Named("p1", "xxx1"),sql.Named("p3", sql.Out{Dest: &tag_id})); err != nil {
fmt.Println(err)
}

now it doesn't work alone

**Comment by sijms** [comment]
this is a complete working code that used named parameter with return clause
```golang
package main

import (
	"database/sql"
	"fmt"
	_ "github.com/sijms/go-ora/v2"
	"os"
	"time"
)

// create sequence
func createSeq(conn *sql.DB) error {
	sqlText := `CREATE SEQUENCE GOORA_TEMP_VISIT_SEQ 
		MINVALUE 1 MAXVALUE 999 
		INCREMENT BY 1 
		START WITH 1 
		NOCACHE  NOCYCLE`
	t := time.Now()
	_, err := conn.Exec(sqlText)
	if err != nil {
		return err
	}
	fmt.Println("Finish create sequence: ", time.Now().Sub(t))
	return nil
}

func createTable(conn *sql.DB) error {
	t := time.Now()
	sqlText := `CREATE TABLE GOORA_TEMP_VISIT(
	VISIT_ID	number(10)	NOT NULL,
	NAME		VARCHAR(200),
	VAL			number(10,2),
	VISIT_DATE	date,
	PRIMARY KEY(VISIT_ID)
	)`
	_, err := conn.Exec(sqlText)
	if err != nil {
		return err
	}
	fmt.Println("Finish create table GOORA_TEMP_VISIT :", time.Now().Sub(t))
	return nil
}

func createTrigger(conn *sql.DB) error {
	sqlText := `CREATE OR REPLACE TRIGGER GOORA_TEMP_VISIT_TRG
BEFORE INSERT ON GOORA_TEMP_VISIT 
FOR EACH ROW 
BEGIN
	SELECT GOORA_TEMP_VISIT_SEQ.NEXTVAL INTO :NEW.VISIT_ID FROM DUAL;
END;`
	t := time.Now()
	_, err := conn.Exec(sqlText)
	if err != nil {
		return err
	}
	fmt.Println("Finish create trigger: ", time.Now().Sub(t))
	return nil
}

func dropTable(conn *sql.DB) error {
	t := time.Now()
	_, err := conn.Exec("drop table GOORA_TEMP_VISIT purge")
	if err != nil {
		return err
	}
	fmt.Println("Finish drop table: ", time.Now().Sub(t))
	return nil
}

func dropSeq(conn *sql.DB) error {
	t := time.Now()
	_, err := conn.Exec("drop sequence GOORA_TEMP_VISIT_SEQ")
	if err != nil {
		return err
	}
	fmt.Println("Finish drop sequence: ", time.Now().Sub(t))
	return nil
}

func dropTrigger(conn *sql.DB) error {
	t := time.Now()
	_, err := conn.Exec("drop trigger GOORA_TEMP_VISIT_TRG")
	if err != nil {
		return err
	}
	fmt.Println("Finish drop trigger: ", time.Now().Sub(t))
	return nil
}

func insertData(conn *sql.DB) error {
	t := time.Now()
        // don't use :DATE as it will produce error so I use :DAT
	stmt, err := conn.Prepare(`INSERT INTO GOORA_TEMP_VISIT(NAME, VAL, VISIT_DATE) 
VALUES(:NAME, :VAL, :DAT) RETURNING VISIT_ID INTO :ID`)
	if err != nil {
		return err
	}
	defer func() {
		_ = stmt.Close()
	}()
	nameText := "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ"
	val := 1.1
	var id int64
	for index := 0; index < 100; index++ {
		_, err = stmt.Exec(sql.Named("NAME", nameText),
			sql.Named("VAL", val),
			sql.Named("DAT", time.Now()),
			sql.Named("ID", sql.Out{Dest: &id}))
		if err != nil {
			return err
		}
		fmt.Printf("Insert row# %d and returning id value = %d\n", index, id)
		id = 3
	}
	fmt.Println("Finish insert data: ", time.Now().Sub(t))
	return nil
}

func main() {
	server := os.Getenv("DSN")
	connStr := os.ExpandEnv(server)
	if connStr == "" {
		fmt.Println("Missing DSN env var")
		os.Exit(1)
	}
	fmt.Println("Connection string: ", connStr)
	conn, err := sql.Open("oracle", server)
	if err != nil {
		fmt.Println("Can't open the driver: ", err)
		return
	}
	defer func() {
		err = conn.Close()
		if err != nil {
			fmt.Println("Can't close driver: ", err)
		}
	}()

	err = conn.Ping()
	if err != nil {
		fmt.Println("Can't ping connection: ", err)
		return
	}

	err = createSeq(conn)
	if err != nil {
		fmt.Println("Can't create sequence", err)
		return
	}
	defer func() {
		err = dropSeq(conn)
		if err != nil {
			fmt.Println("Can't drop sequence: ", err)
		}
	}()

	err = createTable(conn)
	if err != nil {
		fmt.Println("Can't create table: ", err)
		return
	}
	defer func() {
		err = dropTable(conn)
		if err != nil {
			fmt.Println("Can't drop table: ", err)
		}
	}()

	err = createTrigger(conn)
	if err != nil {
		fmt.Println("Can't create trigger", err)
		return
	}
	defer func() {
		err = dropTrigger(conn)
		if err != nil {
			fmt.Println("Can't drop trigger", err)
		}
	}()
	err = insertData(conn)
	if err != nil {
		fmt.Println("Can't insert data", err)
	}
}
```

**Reply by aoticombr** [reply]
the error happens when you only have the :id of return as a parameter, the rest is informed as text.

important detail the error does not happen when you have at least 1 input and an output, if you have only one output

insert into table(column1) values('xxx') returning id into :id

the error happens

however if I run the same query in a block

begin
insert into table(column1) values('xxx') returning id into :id;
end;

the error no longer happens

**Reply by aoticombr** [reply]
**error happens with this example**

db, err := sql.Open("oracle", DB.GetUrlOra())
	if err != nil {
		log.Fatal(err)
	}
	defer db.Close()

	var id_marca int
	//sqlStatement := `BEGIN INSERT INTO MARCAS(MARCAS,ATIVO) VALUES('xxx1','S') RETURNING CDMARCAS INTO :p3; END;`
	sqlStatement := `INSERT INTO MARCAS(MARCAS,ATIVO) VALUES('xxx2','S') RETURNING CDMARCAS INTO :p3`
	stmt, err := db.PrepareContext(context.Background(), sqlStatement)
	if err != nil {
		fmt.Println(err)
	}
	defer stmt.Close()
	if _, err := stmt.ExecContext(context.Background(),
		sql.Named("p3", sql.Out{Dest: &id_marca}),
	); err != nil {
		fmt.Println(err)
	}

**error does not happen if I put an input, even if this input does not exist in the query**


db, err := sql.Open("oracle", DB.GetUrlOra())
	if err != nil {
		log.Fatal(err)
	}
	defer db.Close()

	var id_marca int
	//sqlStatement := `BEGIN INSERT INTO MARCAS(MARCAS,ATIVO) VALUES('xxx1','S') RETURNING CDMARCAS INTO :p3; END;`
	sqlStatement := `INSERT INTO MARCAS(MARCAS,ATIVO) VALUES('xxx2','S') RETURNING CDMARCAS INTO :p3`
	stmt, err := db.PrepareContext(context.Background(), sqlStatement)
	if err != nil {
		fmt.Println(err)
	}
	defer stmt.Close()
	if _, err := stmt.ExecContext(context.Background(),
		sql.Named("p3", sql.Out{Dest: &id_marca}),
		**sql.Named("p1", "xxx1"),**
	); err != nil {
		fmt.Println(err)
	}

**Reply by aoticombr** [reply]
ORA-03111: break received on communication channel
https://www.youtube.com/watch?v=Fonc9rYgZ_o

**Reply by aoticombr** [reply]
simulate erro

func insertData(conn *sql.DB) error {
	t := time.Now()
	// don't use :DATE as it will produce error so I use :DAT
	stmt, err := conn.Prepare(`INSERT INTO GOORA_TEMP_VISIT(NAME, VAL, VISIT_DATE) 
VALUES('tss', 10, sysdate) RETURNING VISIT_ID INTO :ID`)
	if err != nil {
		return err
	}
	defer func() {
		_ = stmt.Close()
	}()
	//nameText := "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ"
	//val := 1.1
	var id int64
	for index := 0; index < 100; index++ {
		_, err = stmt.Exec(
			//	sql.Named("NAME", nameText),
			//	sql.Named("VAL", val),
			//	sql.Named("DAT", time.Now()),
			sql.Named("ID", sql.Out{Dest: &id}))
		if err != nil {
			return err
		}
		fmt.Printf("Insert row# %d and returning id value = %d\n", index, id)
		id = 3
	}
	fmt.Println("Finish insert data: ", time.Now().Sub(t))
	return nil
}

**Reply by aoticombr** [reply]
solution
} else if strings.HasPrefix(uCmdText, "INSERT") ||
		strings.HasPrefix(uCmdText, "MERGE") {
		ret.stmtType = DML
		//ret.bulkExec = true
		ret.bulkExec = false //solution ;)

**Comment by sijms** [comment, marked as answer]
fixed in 2.7.11

**Reply by aoticombr** [reply]
thank ;)
